ext/drreg/drreg.c

```diff
@@ -1830,25 +2856,51 @@ get_tls_data(void *drcontext)
 }
 
 static void
-tls_data_init(per_thread_t *pt)
+tls_data_init(void *drcontext, per_thread_t *pt)
 {
     reg_id_t reg;
     memset(pt, 0, sizeof(*pt));
     for (reg = DR_REG_START_GPR; reg <= DR_REG_STOP_GPR; reg++) {
         drvector_init(&pt->reg[GPR_IDX(reg)].live, 20, false /*!synch*/, NULL);
         pt->reg[GPR_IDX(reg)].native = true;
     }
+#ifdef SIMD_SUPPORTED
+    for (reg = DR_REG_APPLICABLE_START_SIMD; reg <= DR_REG_APPLICABLE_STOP_SIMD; reg++) {
+        drvector_init(&pt->simd_reg[SIMD_IDX(reg)].live, DR_NUM_SIMD_VECTOR_REGS,
+                      false /*!synch*/, NULL);
+        pt->simd_reg[SIMD_IDX(reg)].native = true;
+    }
+    /* We align the block on a 64-byte boundary. */
+    if (drcontext == GLOBAL_DCONTEXT) {
+        pt->simd_spill_start = dr_global_alloc((SIMD_REG_SIZE * MAX_SIMD_SPILLS) + 63);
```

Comment:
I wonder if we should add num_simd_spill_slots to the options, so the tool can decide how much space we allocate, with the same description and requirement of needing another slot per spilled cross-app register. WDYT?

Reply:
Yes, good idea.

Reply:
Wait, didn't we avoid this because we then need to perform reallocs and therefore synch threads?

Reply:
https://github.com/DynamoRIO/dynamorio/issues/3844#issuecomment-534283528

Reply:
On the bases of further investigation of drreg code, and that all multiple calls to drreg_init stem from dr_client_main, I think what you are suggesting is actually doable. Do you agree?

Reply:
On the one hand, the indirection makes it no longer be a severely limited resource, like raw TLS slots are on Windows and Mac.  Thus we can freely allocate a large-ish amount.  OTOH, per-thread space does add up: 32 slots @ 512 bytes each is 16K per thread, so an app with 1K threads (yes not uncommon for us) eats 16M just for these slots, which is significant if they're in the reachable heap.  I guess these aren't reachable: still that's not an insignificant amount of memory.  Shrinking it substantially by letting the client say it only needs 2 or 3 might be worthwhile.

Reply:
Initially, we discussed supporting custom sizes, passing the number of slots required via drreg options. However, the discussion led to us to say that this is a bit difficult due to reallocation and thread syncs. Reading these comments again, I cannot see why these are issues or even why reallocations/thread are necessary. Any idea what was the rationale behind that discussion or were we on a different tangent there.

Reply:
Thanks. Reading this comment I would agree and say maybe static allocation is wiser. Do we really need to support DR_NUM_SIMD_VECTOR_REGS spill slots? I think we can reduce this significantly and document that e.g. we're only supporting up to 4 spills or sth. like that, making the static structure size 8.

Reply:
Perhaps I am missing something trivial here, but the thing that is confusing me is that in the original code of drreg, no thread syncs are needed when increasing the number of raw thread local slots. drreg_init simply frees the current number of slots, then allocates a larger set. Why can't we track the number of SIMD slots in drreg_init also, and use that number in thread_init when performing the allocation. Can drreg_init be called after thread creation? If this is the case, doesn't the allocation of raw thread slots also have a problem which isn't addressed in the current code?

Reply:
TBH I don't really agree with the employment of a hard limit. if you think this is a must, perhaps we should support both approaches.

Reply:
Configurable would be best. @derekbruening , why are you saying we need a full sync? Isn't this per-thread alloc and the ptr gets assigned to tls?

Reply:
> @derekbruening , why are you saying we need a full sync?

Where did I say that?

I seem to recall a completely separate conversation on how drreg_init is not safe and it should document and check that it is only called at process init time: it seems to assume that without documenting or enforcing it.

Reply:
Like I said, I think we were on a different tangent with regards to our discussion. The link in question is here https://github.com/DynamoRIO/dynamorio/issues/3844#issuecomment-534283528

Reply:
Nevertheless the assumption that drreg is initialised at process init time, implies that no threads will be created prior, hence allowing the user the safely specify the number of SIMD register slots prior thread-local allocation.

Reply:
Please add a parameter to the options num_simd_spill_slots and use that to allocate. Then please document that the current limitation both for num_spill_slots and num_simd_spill_slots changed by drreg_init() is that this happens only during init time. You could even file an issue and reference with XXX or FIXME. @derekbruening ok with that?

Reply:
There is a further restriction: drreg operations that use spill slots can't be used until all components have called drreg_init.  This could be problematic for clients/extensions who want to use drreg for some static code tweaking during process init.  I don't think we have any existing code like that in the DR repo, but Dr. Memory might do that when converted to drreg (WIP branch https://github.com/DynamoRIO/drmemory/tree/i1795-use-drreg): would have to search.  It has out-of-line calls, with an array of them for each of different registers to restore.  Any ideas for solving that?

Reply:
Hmm, looks like I enabled use during process init for DrM: #2910.

Reply:
Filed #3934

Reply:
Done